medium.com/InterfaceInGo-Part2.md

```diff
@@ -0,0 +1,619 @@
+}
+```
+
+编译的输出：
+
+```golang
+main.go:16: cannot convert v (type I1) to type I2:
+	I1 does not implement I2 (missing N method)
+```
+
+原因和之前一样，如果`I2`的方法集是`I1`方法集的子集，编译器将会在编译阶段知道这个。但是不同的是，这种转换只有在运行时才可以进行。
+
+> 这不是严格意义上的类型转换，而是类型断言和类型切换，允许检查/检索接口类型值的动态值或者将接口类型的值转换为不同接口类型的值。
+
+## 类型断言
+
+类型断言的语法格式如下：
+
+```golang
+v.(T)
+```
+
+其中v为接口类型，T为抽象类型或具体值类型
+
+### 具体类型
+首先来让我们看一下，非接口类型（non-interface）是如何工作的（代码如下）：
+
+```golang
+type I interface {
+    M()
+}
+type T struct{}
+
+func (T) M() {}
+
+func main() {
+    var v1 I = T{}
+    v2 := v1.(T)
+    fmt.Printf("%T\n", v2) // main.T
+}
+```
+
+类型断言指定的类型必须实现变量`v1`的接口类型`I`，这将在编译阶段被验证（代码如下）：
+
+```golang
+type I interface {
+    M()
+}
+
+type T1 struct{}
+
+func (T1) M() {}
+
+type T2 struct{}
+
+func main() {
+    var v1 I = T1{}
+    v2 := v1.(T2)
+    fmt.Printf("%T\n", v2)
+}
+```
+
+成功编译这样的代码是不可能的，因为会抛出这个错误`impossible type assertion`，变量`v1`不能持有任何类型`T2`的值，因为`T2`不满足接口`I`同时变量`v1`只能存储实现了接口`I`的类型的值。
+
+在程序运行的过程中，编译器不知道变量`v1`存储了什么类型的值。类型断言是一种可以检索接口类型值的动态值的方法。但是如果`v1`的动态类型和`T`不匹配的时候会发生什么呢？（代码如下）：
+
+```golang
+type I interface {
+    M()
+}
+type T1 struct{}
+
+func (T1) M() {}
+
+type T2 struct{}
+
+func (T2) M() {}
+
+func main() {
+    var v1 I = T1{}
+    v2 := v1.(T2)
+    fmt.Printf("%T\n", v2)
+}
+```
+
+程序将会不知所措（panic）：
+
+```golang
+panic: interface conversion: main.I is main.T1, not main.T2
+```
+
+### 多值转化（请不要惊慌）
+
+类型断言可以以多值的形式使用，附加的第二个值是一个boolean值，表明断言是否成立。如果不成立，第一个值是一个`T`类型的零值（zero-value）。（代码如下）：
+
+```golang
+type I interface {
+    M()
+}
+
+type T1 struct{}
+
+func (T1) M() {}
+
+type T2 struct{}
+
+func (T2) M() {}
+
+func main() {
+    var v1 I = T1{}
+    v2, ok := v1.(T2)
+    if !ok {
+        fmt.Printf("ok: %v\n", ok) // ok: false
+        fmt.Printf("%v,  %T\n", v2, v2) // {},  main.T2
+    }
+}
+```
+
+这种形式不会引起恐慌，因为返回的第二个布尔类型的值可以用来检查断言是否成立。
+
+### 接口类型
+
+在大多数情况下使用类型断言是没问题的。golang还允许传递接口类型。它会检查动态类型是否满足接口的要求，并且返回该接口类型值的值。在转换条约中，传递给类型断言的接口方法集合不用必须是变量`v`类型的子集（代码如下）：
+
+```golang
+type I1 interface {
+    M()
+}
+
+type I2 interface {
+    I1
+    N()
+}
+
+type T struct{
+    name string
+}
+
+func (T) M() {}
+
+func (T) N() {}
+
+func main() {
+    var v1 I1 = T{"foo"}
+    var v2 I2
+    v2, ok := v1.(I2)
+    fmt.Printf("%T %v %v\n", v2, v2, ok) // main.T {foo} true
+}
+```
+
+如果接口不满足，那么会返回接口的`零值（zero-value）`，也就是nil（代码如下）：
+
+```golang
+type I1 interface {
+    M()
+}
+
+type I2 interface {
+    N()
+}
+
+type T struct {}
+
+func (T) M() {}
+
+func main() {
+    var v1 I1 = T{}
+    var v2 I2
+    v2, ok := v1.(I2)
+    fmt.Printf("%T %v %v\n", v2, v2, ok) // <nil> <nil> false
+}
+```
+
+> 当处理接口类型时，还支持类型断言的单值转换。
+
+### nil
+
+当变量`v`是nil，那么类型断言将总是false。不论`T`是一个接口类型还是一个具体类型（代码如下）：
+
+```golang
+type I interface {
+    M()
+}
+
+type T struct{}
+
+func (T) M() {}
+
+func main() {
+    var v1 I
+    v2 := v1.(T)
+    fmt.Printf("%T\n", v2)
+}
+```
+
+当开始运行这段程序时，程序将会恐慌。
+
+```golang
+panic: interface conversion: main.I is nil, not main.T
+```
+
+前面介绍的多值类型则可以当`v`是nil的时候，避免出现恐慌(panic)。
```

Comment:
"panic" is a professional vocabulary and does not require translation

Reply:
yeah、may be its means just like '程序崩溃'， but can't be translated to '恐慌'.

Reply:
so i think， we can translate panic as '程序崩溃'、'挂掉'、'宕机' and so on...;

Reply:
"panic" just like "Expection", it is a keyword  of golang。 you can use "panic" in you source code, it will make you program crash. so I think,  we need't translated it. How do you think about